exporters/class-exporter-csv.php

```diff
@@ -38,12 +38,18 @@ public function output_file( $data, $columns ) {
 			header( 'Content-Disposition: attachment; filename="stream.csv"' );
 		}
 
-		$output = join( ',', array_values( $columns ) ) . "\n";
+		ob_start();
+
+		$csv = fopen( 'php://output', 'w' );
+		fputcsv( $csv, array_values( $columns ) );
+
 		foreach ( $data as $row ) {
-			$output .= join( ',', $row ) . "\n";
+			fputcsv( $csv, $row );
 		}
+		fclose( $csv ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_fclose
+
+		echo ob_get_clean(); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```

Comment:
@delawski I'm worried about the size of the output buffer, do you think I need to be?

Reply:
See https://github.com/xwp/stream/issues/1476 and https://github.com/xwp/stream/issues/1503 for some large log sizes

Reply:
That's a good question. I guess that with a high enough number of records we may hit a PHP timeout limit (depending on the server configuration). However, some pro/enterprise users may still want to export everything.

I think a reasonable short-term solution to that issue would be to allow setting the number of records to export (as suggested [here](https://github.com/xwp/stream/issues/1476#issuecomment-2034374922)). We could default to 10000 records but make that value filterable.

A more sophisticated solution (maybe for the future) would be to rely on some async queue (Action Scheduler?) that would export the records to a file stored in the filesystem. Once everything is exported, a link to that file would be available in WP admin.

What are your thoughts?

Reply:
> I think a reasonable short-term solution to that issue would be to allow setting the number of records to export (as suggested https://github.com/xwp/stream/issues/1476#issuecomment-2034374922). We could default to 10000 records but make that value filterable.

I'm not sure how to handle pagination though? I guess a first step is allowing that to be filterable and then figuring out if / where people are hitting the php limits (and testing that ourselves!)

> A more sophisticated solution (maybe for the future) would be to rely on some async queue (Action Scheduler?) that would export the records to a file stored in the filesystem. Once everything is exported, a link to that file would be available in WP admin.

I like this, it should be straightforward although we'd need to think about security but we could also make it be deleted after x amount of time too. (I 💖 Action Scheduler)

Reply:
@delawski Yes, that makes sense. There's an [open issue](https://github.com/xwp/stream/issues/1476) for filtering the number. The WP CLI command has a `records_per_page` field that can be used to get an arbitrary number of records.